src/dev.js

```diff
@@ -0,0 +1,221 @@
+// assumptions and create a bounding box of this size.
+const presumedHeight = 25;
+const presumedWidth = 400;
+
+function storeTagPosition(id, top, left) {
+  const exists = tagPositions.some(t => t.id === id);
+  const position = {
+    id,
+    top,
+    left,
+    bottom: top + presumedHeight,
+    right: left + presumedWidth,
+  };
+
+  if (!exists) {
+    tagPositions.push(position);
+  } else {
+    const index = tagPositions.findIndex(t => t.id === id);
+    tagPositions[index] = {
+      ...position,
+    };
+  }
+}
+
+function removeTagPosition(id) {
+  const index = tagPositions.findIndex(t => t.id === id);
+  if (index !== -1) {
+    tagPositions.splice(index, 1);
+  }
+}
+
+// Returns true if the proposed location (with the height/width presumptions) would overlap with another
+// tag for a different ID.
+function overlapExists(id, refTop, refLeft) {
+  const provisionalBottom = refTop + presumedHeight;
+  const provisionalRight = refLeft + presumedWidth;
+
+  return tagPositions.some(t => {
+    if (t.id === id) return false;
+    if (provisionalRight < t.left || t.right < refLeft) {
+      return false;
+    }
+    if (provisionalBottom < t.top || t.bottom < refTop) {
+      return false;
+    }
+    return true;
+  });
+}
+
+function getNewPosition(id, refTop, refLeft) {
+  let top = refTop;
+  while (overlapExists(id, top, refLeft)) {
+    // Keep moving the tag location downwards until we no longer have an overlap.
+    top++;
+  }
+
+  return {top, left: refLeft};
+}
+
+function IDTag({targetRef, id}) {
+  const [position, setPosition] = useState(null);
+  const [copied, setCopied] = useState(false);
+
+  useEffect(() => {
+    const node = targetRef.current;
+    if (!node) return;
+
+    let frameId;
+    const animationFrameHandler = () => {
+      const rect = node.getBoundingClientRect();
+      const parentRect = node.parentElement.getBoundingClientRect();
+
+      if (parentRect.width === 0 || parentRect.height === 0) {
+        // If width/height is 0 then the element is not visible so we should avoid showing the tag.
+        setPosition(null);
+      } else {
+        const newPosition = getNewPosition(
+          id,
+          rect.top + window.scrollY,
+          rect.left + window.scrollX
+        );
+        storeTagPosition(id, newPosition.top, newPosition.left);
+        setPosition(newPosition);
+      }
+
+      // Poll this function constantly without affecting UI performance
+      frameId = requestAnimationFrame(animationFrameHandler);
+    };
+
+    animationFrameHandler();
+
+    return () => {
+      removeTagPosition(id);
+      cancelAnimationFrame(frameId);
+    };
+  }, [targetRef, id]);
+
+  const onClick = useCallback(() => {
+    navigator.clipboard.writeText(id);
+    setCopied(true);
+    setTimeout(() => setCopied(false), 1000);
+  }, [id]);
+
+  if (!position) return null;
+
+  return (
+    <button
+      type="button"
+      onClick={onClick}
+      style={{
```

Comment:
very minor since it is meant to be used in development mode: I would suggest checking how inline styling plays out with more restrictive content security policy that does not allow inline styling

Reply:
Ah I did not even think about this, I can have a look into it :)

Reply:
From taking a brief look at this, Zenodo and CDS as well as the default cookie cutter config would not cause inline styles to be rejected via the CSP header. However, it's possible that some instances may have custom headers that are more strict. I suppose it's a tradeoff between adding complexity via a PostCSS plugin for rollup in this repo VS adding complexity by requiring users to allow inline styles in any custom CSP they write. I'm not sure what's the best choice here!

Reply:
I don't think this is a problem at all in halfway modern React versions: https://github.com/facebook/react/issues/5878

Reply:
> Also, how likely is it that people use this debug mode on a production system in the first place?

unlikely, this is why I believe it is very minor, almost nitpick :)